File: tools/build_estacionalidad_workbook.py
from __future__ import annotations

import json
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
from openpyxl.worksheet.table import Table, TableStyleInfo
from openpyxl.utils import get_column_letter


OUT_DIR = Path(r"C:\Users\chval\OneDrive\Escritorio\PLATAFORMA GESTION\outputs\estacionalidad_consolidado")
INPUT = OUT_DIR / "estacionalidad_data.json"
OUTPUT = OUT_DIR / "Historico Estacionalidad HYL.xlsx"

BLUE = "1F4E79"
DARK_BLUE = "17365D"
LIGHT_BLUE = "D9EAF7"
LIGHT_GRAY = "F3F6F8"
BORDER = Side(style="thin", color="D9E2F3")


def value(row, key):
    val = row.get(key)
    return "" if val is None else val


def add_table(ws, name, rows, columns):
    ws.append([label for _, label, _ in columns])
    for row in rows:
        ws.append([value(row, key) for key, _, _ in columns])

    max_row = max(ws.max_row, 2)
    max_col = len(columns)
    ref = f"A1:{get_column_letter(max_col)}{max_row}"

    table = Table(displayName=name, ref=ref)
    style = TableStyleInfo(
        name="TableStyleMedium2",
        showFirstColumn=False,
        showLastColumn=False,
        showRowStripes=True,
        showColumnStripes=False,
    )
    table.tableStyleInfo = style
    ws.add_table(table)
    ws.freeze_panes = "A2"
    ws.auto_filter.ref = ref
    ws.sheet_view.showGridLines = False

    for cell in ws[1]:
        cell.fill = PatternFill("solid", fgColor=BLUE)
        cell.font = Font(color="FFFFFF", bold=True)
        cell.alignment = Alignment(horizontal="center")

    for col_idx, (_, _, number_format) in enumerate(columns, start=1):
        letter = get_column_letter(col_idx)
        width = 12
        for cell in ws[letter]:
            text = str(cell.value or "")
            width = min(max(width, len(text) + 2), 55)
            if cell.row > 1 and number_format:
                cell.number_format = number_format
        ws.column_dimensions[letter].width = width

    for row in ws.iter_rows():
        for cell in row:
            cell.border = Border(bottom=BORDER)
            cell.alignment = Alignment(vertical="top", wrap_text=False)


def build_summary(wb, data):
    ws = wb.active
    ws.title = "Resumen"
    ws.sheet_view.showGridLines = False
    ws.merge_cells("A1:H1")
    ws["A1"] = "Historico de estacionalidad H&L GYM"
    ws["A1"].fill = PatternFill("solid", fgColor=DARK_BLUE)
    ws["A1"].font = Font(color="FFFFFF", bold=True, size=16)
    ws["A1"].alignment = Alignment(horizontal="center")

    ws["A3"] = "Metrica"
    ws["B3"] = "Valor"
    for col in ("A", "B"):
        ws[f"{col}3"].fill = PatternFill("solid", fgColor=BLUE)
        ws[f"{col}3"].font = Font(color="FFFFFF", bold=True)
    for index, item in enumerate(data["resumen"], start=4):
        ws[f"A{index}"] = item["metrica"]
        ws[f"B{index}"] = item["valor"]

    for row in ws["A3:B9"]:
        for cell in row:
            cell.border = Border(bottom=BORDER, top=BORDER, left=BORDER, right=BORDER)

    notes = [
        "Uso recomendado",
        "Historico gerencial agregado: metas, ejecutado, ranking y usuarios activos.",
        "No reemplaza un historico transaccional de ventas por cliente/factura.",
        "Los nombres fueron normalizados, pero se conserva el valor original y archivo fuente.",
        "Los temporales de Excel ~$ fueron ignorados.",
        "En usuarios activos se estima la fecha usando el mes del archivo cuando la etiqueta del dia no es confiable.",
    ]
    ws.merge_cells("D3:H3")
    ws["D3"] = notes[0]
    ws["D3"].fill = PatternFill("solid", fgColor=BLUE)
    ws["D3"].font = Font(color="FFFFFF", bold=True)
    for idx, note in enumerate(notes[1:], start=4):
        ws.merge_cells(start_row=idx, start_column=4, end_row=idx, end_column=8)
        cell = ws.cell(idx, 4, note)
        cell.fill = PatternFill("solid", fgColor=LIGHT_GRAY)
        cell.alignment = Alignment(wrap_text=True, vertical="top")

    widths = {"A": 28, "B": 22, "D": 28, "E": 18, "F": 18, "G": 18, "H": 18}
    for col, width in widths.items():
        ws.column_dimensions[col].width = width
    for row in range(4, 9):
        ws.row_dimensions[row].height = 34


def main():
    data = json.loads(INPUT.read_text(encoding="utf-8"))
    wb = Workbook()
    build_summary(wb, data)

    cobertura_cols = [
        ("periodo", "Periodo", None),
        ("anio", "Anio", None),
        ("mes_num", "Mes Num", None),
        ("mes", "Mes", None),
        ("archivo", "Archivo", None),
        ("hojas", "Hojas", None),
        ("tiene_dia_a_dia", "Tiene Dia a Dia", None),
        ("tiene_ranking_sede", "Tiene Ranking Sede", None),
        ("registros_ranking_sede", "Reg Ranking Sede", None),
        ("tiene_ranking_asesores", "Tiene Ranking Asesores", None),
        ("registros_ranking_asesores", "Reg Ranking Asesores", None),
        ("tiene_usuarios_activos", "Tiene Usuarios Activos", None),
        ("registros_usuarios_activos", "Reg Usuarios Activos", None),
        ("fuente_archivo", "Fuente Archivo", None),
    ]
    ws = wb.create_sheet("Cobertura")
    add_table(ws, "tblCobertura", data["cobertura"], cobertura_cols)

    rank_cols = [
        ("periodo", "Periodo", None),
        ("anio", "Anio", None),
        ("mes_num", "Mes Num", None),
        ("mes", "Mes", None),
        ("entidad_normalizada", "Entidad Normalizada", None),
        ("entidad_original", "Entidad Original", None),
        ("puesto", "Puesto", None),
        ("ejecutado_total", "Ejecutado Total", "#,##0"),
        ("presupuesto", "Presupuesto", "#,##0"),
        ("meta_1", "Meta 1", "#,##0"),
        ("meta_1_5", "Meta 1.5", "#,##0"),
        ("meta_2", "Meta 2", "#,##0"),
        ("meta_2_5", "Meta 2.5", "#,##0"),
        ("meta_3", "Meta 3", "#,##0"),
        ("meta_4", "Meta 4", "#,##0"),
        ("porcentaje_cumplimiento", "% Cumplimiento", "0.0%"),
        ("porcentaje_a_hoy", "% A Hoy", "0.0%"),
        ("diferencia_porcentaje", "Diferencia %", "0.0%"),
        ("valor_a_hoy", "$ A Hoy", "#,##0"),
        ("hoja", "Hoja", None),
        ("fila_origen", "Fila Origen", None),
        ("formula_ejecutado", "Formula Ejecutado", None),
        ("fuente_archivo", "Fuente Archivo", None),
    ]
    ws = wb.create_sheet("Ranking Sedes")
    add_table(ws, "tblRankingSedes", data["ranking_sedes"], rank_cols)
    ws = wb.create_sheet("Ranking Asesores")
    add_table(ws, "tblRankingAsesores", data["ranking_asesores"], rank_cols)

    activos_cols = [
        ("periodo", "Periodo", None),
        ("anio", "Anio", None),
        ("mes_num", "Mes Num", None),
        ("mes", "Mes", None),
        ("fecha_estimada", "Fecha Estimada", "yyyy-mm-dd"),
        ("dia", "Dia", None),
        ("fecha_original", "Fecha Original", None),
        ("sede_normalizada", "Sede Normalizada", None),
        ("sede_original", "Sede Original", None),
        ("usuarios_activos", "Usuarios Activos", "#,##0"),
        ("hoja", "Hoja", None),
        ("fila_origen", "Fila Origen", None),
        ("formula_valor", "Formula Valor", None),
        ("fuente_archivo", "Fuente Archivo", None),
    ]
    ws = wb.create_sheet("Usuarios Activos")
    add_table(ws, "tblUsuariosActivos", data["usuarios_activos"], activos_cols)

    sedes_mes = {}
    for row in data["ranking_sedes"]:
        key = (row["periodo"], row["anio"], row["mes_num"], row["entidad_normalizada"])
        item = sedes_mes.setdefault(
            key,
            {
                "periodo": row["periodo"],
                "anio": row["anio"],
                "mes_num": row["mes_num"],
                "sede": row["entidad_normalizada"],
                "ejecutado_total": 0,
                "registros": 0,
            },
        )
        item["ejecutado_total"] += row.get("ejecutado_total") or 0
        item["registros"] += 1
    sedes_mes_rows = sorted(sedes_mes.values(), key=lambda item: (item["periodo"], item["sede"]))
    sedes_mes_cols = [
        ("periodo", "Periodo", None),
        ("anio", "Anio", None),
        ("mes_num", "Mes Num", None),
        ("sede", "Sede", None),
        ("ejecutado_total", "Ejecutado Total", "#,##0"),
        ("registros", "Registros", None),
    ]
    ws = wb.create_sheet("Sedes Mes")
    add_table(ws, "tblSedesMes", sedes_mes_rows, sedes_mes_cols)

    wb.save(OUTPUT)
    print(f"OK {OUTPUT}")
    print(f"Cobertura: {len(data['cobertura'])}")
    print(f"Ranking sedes: {len(data['ranking_sedes'])}")
    print(f"Ranking asesores: {len(data['ranking_asesores'])}")
    print(f"Usuarios activos: {len(data['usuarios_activos'])}")


if __name__ == "__main__":
    main()

File: tools/build_estacionalidad_workbook.mjs
import fs from "node:fs/promises";
import { SpreadsheetFile, Workbook } from "@oai/artifact-tool";

const outputDir = "C:/Users/chval/OneDrive/Escritorio/PLATAFORMA GESTION/outputs/estacionalidad_consolidado";
const inputPath = `${outputDir}/estacionalidad_data.json`;
const outputPath = `${outputDir}/Historico Estacionalidad HYL.xlsx`;

const data = JSON.parse(await fs.readFile(inputPath, "utf8"));

const workbook = Workbook.create();

function aoaFromObjects(rows, columns) {
  return [
    columns.map((column) => column.label),
    ...rows.map((row) => columns.map((column) => {
      const value = row[column.key];
      return value === undefined ? null : value;
    })),
  ];
}

function colLetter(index) {
  let n = index + 1;
  let out = "";
  while (n > 0) {
    const rem = (n - 1) % 26;
    out = String.fromCharCode(65 + rem) + out;
    n = Math.floor((n - 1) / 26);
  }
  return out;
}

function writeTable(sheet, startCell, name, rows, columns) {
  const values = aoaFromObjects(rows, columns);
  const rowCount = Math.max(values.length, 2);
  const colCount = columns.length;
  const range = sheet.getRangeByIndexes(0, 0, rowCount, colCount);
  range.values = values.length === 1 ? [...values, columns.map(() => null)] : values;
  const tableRange = `A1:${colLetter(colCount - 1)}${rowCount}`;
  const table = sheet.tables.add(tableRange, true, name);
  table.style = "TableStyleMedium2";
  sheet.freezePanes.freezeRows(1);
  sheet.showGridLines = false;
  sheet.getRange(`A1:${colLetter(colCount - 1)}1`).format = {
    fill: "#1F4E79",
    font: { bold: true, color: "#FFFFFF" },
  };
  sheet.getRange(tableRange).format.autofitColumns();
  sheet.getRange(tableRange).format.autofitRows();
  return table;
}

function applyNumberFormats(sheet, columns, rowCount) {
  if (rowCount < 2) return;
  columns.forEach((column, index) => {
    if (!column.format) return;
    const letter = colLetter(index);
    sheet.getRange(`${letter}2:${letter}${rowCount}`).format.numberFormat = column.format;
  });
}

function summarizeByPeriod(rows, entityKey, valueKey) {
  const map = new Map();
  for (const row of rows) {
    const key = `${row.periodo}|${row[entityKey] || ""}`;
    if (!map.has(key)) {
      map.set(key, {
        periodo: row.periodo,
        anio: row.anio,
        mes_num: row.mes_num,
        entidad: row[entityKey] || "",
        ejecutado_total: 0,
        registros: 0,
      });
    }
    const item = map.get(key);
    item.ejecutado_total += Number(row[valueKey] || 0);
    item.registros += 1;
  }
  return Array.from(map.values()).sort((a, b) => `${a.periodo}${a.entidad}`.localeCompare(`${b.periodo}${b.entidad}`));
}

const summary = workbook.worksheets.add("Resumen");
summary.showGridLines = false;
summary.getRange("A1:H1").merge();
summary.getRange("A1").values = [["Historico de estacionalidad H&L GYM"]];
summary.getRange("A1").format = {
  fill: "#17365D",
  font: { bold: true, color: "#FFFFFF", size: 16 },
};
summary.getRange("A3:B8").values = [
  ["Metrica", "Valor"],
  ...data.resumen.map((item) => [item.metrica, item.valor]),
];
summary.getRange("A3:B3").format = {
  fill: "#1F4E79",
  font: { bold: true, color: "#FFFFFF" },
};
summary.getRange("A3:B8").format.borders = { preset: "all", style: "thin", color: "#D9E2F3" };
summary.getRange("D3:H10").values = [
  ["Uso recomendado", null, null, null, null],
  ["Historico gerencial agregado: metas, ejecutado, ranking y usuarios activos.", null, null, null, null],
  ["No reemplaza un historico transaccional de ventas por cliente/factura.", null, null, null, null],
  ["Los nombres fueron normalizados, pero se conserva el valor original y archivo fuente.", null, null, null, null],
  ["Los temporales de Excel ~$ fueron ignorados.", null, null, null, null],
  ["En usuarios activos se estima la fecha usando el mes del archivo cuando la etiqueta del dia no es confiable.", null, null, null, null],
  [null, null, null, null, null],
  ["Hojas incluidas: Cobertura, Ranking Sedes, Ranking Asesores, Usuarios Activos, Sedes Mes.", null, null, null, null],
];
summary.getRange("D3:H3").merge();
summary.getRange("D4:H10").merge(true);
summary.getRange("D3:H10").format = {
  fill: "#F2F6FA",
  font: { color: "#1F2937" },
  wrapText: true,
};
summary.getRange("A3:B8").format.autofitColumns();
summary.getRange("D3:H10").format.rowHeight = 30;

const coberturaCols = [
  { key: "periodo", label: "Periodo" },
  { key: "anio", label: "Anio" },
  { key: "mes_num", label: "Mes Num" },
  { key: "mes", label: "Mes" },
  { key: "archivo", label: "Archivo" },
  { key: "hojas", label: "Hojas" },
  { key: "tiene_dia_a_dia", label: "Tiene Dia a Dia" },
  { key: "tiene_ranking_sede", label: "Tiene Ranking Sede" },
  { key: "registros_ranking_sede", label: "Reg Ranking Sede" },
  { key: "tiene_ranking_asesores", label: "Tiene Ranking Asesores" },
  { key: "registros_ranking_asesores", label: "Reg Ranking Asesores" },
  { key: "tiene_usuarios_activos", label: "Tiene Usuarios Activos" },
  { key: "registros_usuarios_activos", label: "Reg Usuarios Activos" },
  { key: "fuente_archivo", label: "Fuente Archivo" },
];
const cobertura = workbook.worksheets.add("Cobertura");
writeTable(cobertura, "A1", "tblCobertura", data.cobertura, coberturaCols);

const rankCols = [
  { key: "periodo", label: "Periodo" },
  { key: "anio", label: "Anio" },
  { key: "mes_num", label: "Mes Num" },
  { key: "mes", label: "Mes" },
  { key: "entidad_normalizada", label: "Entidad Normalizada" },
  { key: "entidad_original", label: "Entidad Original" },
  { key: "puesto", label: "Puesto" },
  { key: "ejecutado_total", label: "Ejecutado Total", format: "#,##0" },
  { key: "presupuesto", label: "Presupuesto", format: "#,##0" },
  { key: "meta_1", label: "Meta 1", format: "#,##0" },
  { key: "meta_1_5", label: "Meta 1.5", format: "#,##0" },
  { key: "meta_2", label: "Meta 2", format: "#,##0" },
  { key: "meta_2_5", label: "Meta 2.5", format: "#,##0" },
  { key: "meta_3", label: "Meta 3", format: "#,##0" },
  { key: "meta_4", label: "Meta 4", format: "#,##0" },
  { key: "porcentaje_cumplimiento", label: "% Cumplimiento", format: "0.0%" },
  { key: "porcentaje_a_hoy", label: "% A Hoy", format: "0.0%" },
  { key: "diferencia_porcentaje", label: "Diferencia %", format: "0.0%" },
  { key: "valor_a_hoy", label: "$ A Hoy", format: "#,##0" },
  { key: "hoja", label: "Hoja" },
  { key: "fila_origen", label: "Fila Origen" },
  { key: "formula_ejecutado", label: "Formula Ejecutado" },
  { key: "fuente_archivo", label: "Fuente Archivo" },
];

const rankSedes = workbook.worksheets.add("Ranking Sedes");
writeTable(rankSedes, "A1", "tblRankingSedes", data.ranking_sedes, rankCols);
applyNumberFormats(rankSedes, rankCols, data.ranking_sedes.length + 1);

const rankAsesores = workbook.worksheets.add("Ranking Asesores");
writeTable(rankAsesores, "A1", "tblRankingAsesores", data.ranking_asesores, rankCols);
applyNumberFormats(rankAsesores, rankCols, data.ranking_asesores.length + 1);

const activosCols = [
  { key: "periodo", label: "Periodo" },
  { key: "anio", label: "Anio" },
  { key: "mes_num", label: "Mes Num" },
  { key: "mes", label: "Mes" },
  { key: "fecha_estimada", label: "Fecha Estimada" },
  { key: "dia", label: "Dia" },
  { key: "fecha_original", label: "Fecha Original" },
  { key: "sede_normalizada", label: "Sede Normalizada" },
  { key: "sede_original", label: "Sede Original" },
  { key: "usuarios_activos", label: "Usuarios Activos", format: "#,##0" },
  { key: "hoja", label: "Hoja" },
  { key: "fila_origen", label: "Fila Origen" },
  { key: "formula_valor", label: "Formula Valor" },
  { key: "fuente_archivo", label: "Fuente Archivo" },
];
const activos = workbook.worksheets.add("Usuarios Activos");
writeTable(activos, "A1", "tblUsuariosActivos", data.usuarios_activos, activosCols);
applyNumberFormats(activos, activosCols, data.usuarios_activos.length + 1);

const sedesMesRows = summarizeByPeriod(data.ranking_sedes, "entidad_normalizada", "ejecutado_total");
const sedesMesCols = [
  { key: "periodo", label: "Periodo" },
  { key: "anio", label: "Anio" },
  { key: "mes_num", label: "Mes Num" },
  { key: "entidad", label: "Sede" },
  { key: "ejecutado_total", label: "Ejecutado Total", format: "#,##0" },
  { key: "registros", label: "Registros" },
];
const sedesMes = workbook.worksheets.add("Sedes Mes");
writeTable(sedesMes, "A1", "tblSedesMes", sedesMesRows, sedesMesCols);
applyNumberFormats(sedesMes, sedesMesCols, sedesMesRows.length + 1);

for (const sheet of [cobertura, rankSedes, rankAsesores, activos, sedesMes]) {
  const used = sheet.getUsedRange();
  used.format.wrapText = false;
}

await fs.mkdir(outputDir, { recursive: true });

const preview = await workbook.render({
  sheetName: "Resumen",
  autoCrop: "all",
  scale: 1,
  format: "png",
});
await fs.writeFile(`${outputDir}/preview_resumen.png`, new Uint8Array(await preview.arrayBuffer()));

const xlsx = await SpreadsheetFile.exportXlsx(workbook);
await xlsx.save(outputPath);

console.log(`OK ${outputPath}`);
console.log(`Cobertura: ${data.cobertura.length}`);
console.log(`Ranking sedes: ${data.ranking_sedes.length}`);
console.log(`Ranking asesores: ${data.ranking_asesores.length}`);
console.log(`Usuarios activos: ${data.usuarios_activos.length}`);
